doc/neps/nep-0056-array-api-main-namespace.rst

```diff
@@ -0,0 +1,524 @@
+denominator. As a concrete example, this code is taken from ``scipy.cluster``:
+
+.. code:: python
+
+    def vq_py(obs, code_book, check_finite=True):
+        """Python version of vq algorithm"""
+        xp = array_namespace(obs, code_book)
+        obs = as_xparray(obs, xp=xp, check_finite=check_finite)
+        code_book = as_xparray(code_book, xp=xp, check_finite=check_finite)
+
+        if obs.ndim != code_book.ndim:
+            raise ValueError("Observation and code_book should have the same rank")
+
+        if obs.ndim == 1:
+            obs = obs[:, xp.newaxis]
+            code_book = code_book[:, xp.newaxis]
+
+        # Once `cdist` has array API support, this `xp.asarray` call can be removed
+        dist = xp.asarray(cdist(obs, code_book))
+        code = xp.argmin(dist, axis=1)
+        min_dist = xp.min(dist, axis=1)
+        return code, min_dist
+
+It mostly looks like normal NumPy code, but will run with for example PyTorch
+tensors as input and then return PyTorch tensors. There is a lot more to this
+story of course then this basic example. These blog posts on scikit-learn [1]_
+and SciPy's [2]_ experiences and impact (large performance gains in some cases
+- ``LinearDiscriminantAnalysis.fit`` showed ~28x gain with PyTorch on GPU vs.
+NumPy) paint a more complete picture.
+
+For end users who are using NumPy directly, little changes aside from there
+being fewer differences between NumPy and other libraries they may want to use
+as well. This shortens their learning curve and makes it easier to switch
+between NumPy and PyTorch/JAX/CuPy. In addition, they should benefit from
+array-consuming libraries starting to support multiple array libraries, making
+their experience of using a stack of Python packages for scientific computing
+or data science more seamless.
+
+Finally, for authors of other array libraries as well as tools like Numba,
+array API standard support should save them time. The design rules ([3]_), and
+in some cases new APIs like the ``unique_*`` ones, are easier to implement on
+GPU and for JIT compilers as a result of more predictable behavior.
+
+
+Backward compatibility
+----------------------
+
+The changes that have a backwards compatibility impact fall into these
+categories:
+
+1. Raising errors for consistency/strictness in some places where NumPy now
+   allows more flexible behavior,
+2. Dtypes of returned arrays for some element-wise functions and reductions,
+3. Numerical behavior for few tolerance keywords,
+4. Functions moved to ``numpy.linalg`` and supporting stacking/batching.
+
+Raising errors for consistency/strictness includes:
+
+1. Making ``.T`` error for >2 dimensions,
+2. Making ``cross`` error on size-2 vectors (only size-3 vectors are supported),
+3. Making ``solve`` error on ambiguous input (only accept ``x2`` as vector if ``x2.ndim == 1``),
+4. ``outer`` raises rather than flattens on >1-D inputs,
+5. In-place operators are disallowed when the left-hand side would be promoted.
```

Comment:
It looks like there is nothing to do here - things are already behaving this way:
```python
>>> # integer dtypes
>>> x = np.ones((2,2), dtype=np.int32)
>>> x += np.ones((2, 2), dtype=np.int64)
>>> x
array([[2, 2],
       [2, 2]], dtype=int32)
>>> x += np.full((2, 2), 2**34, dtype=np.int64)
>>> x
array([[2, 2],
       [2, 2]], dtype=int32)

>>> # floating-point dtypes
>>> x = np.ones((2,2), dtype=np.float32)
>>> x += np.ones((2, 2), dtype=np.int64)
>>> x
array([[2., 2.],
       [2., 2.]], dtype=float32)
>>> x += np.ones((2, 2), dtype=np.float64)
>>> x
array([[3., 3.],
       [3., 3.]], dtype=float32)
>>> x -= x.mean()
>>> x
array([[0., 0.],
       [0., 0.]], dtype=float32)
>>> x -= x.mean().astype(np.float64)
>>> x
array([[0., 0.],
       [0., 0.]], dtype=float32)
```

I think I took this requirement from https://numpy.org/devdocs/reference/array_api.html#type-promotion-differences, but the example given there seems wrong (I checked also against 1.23.5), there is no upcasting. So, false alarm it looks like - sentence can be removed.

Reply:
Now of course there's the opposite issue:
```python
>>> a = np.array(1, dtype=np.int8)
>>> a += np.array(2**12, dtype=np.int16)
>>> a
array(1, dtype=int8)
>>> a + np.array(2**12, dtype=np.int16)
4097
```
It looks like when the right-hand side has a higher-precision dtype, what happens is that it gets cast to the lower-precision dtype silently. Which results in all zeros for values that exceed the max value of the left-hand side dtype. And hence a wrong result. So `x += y` is not equal to `x = x + y` at that point. It would not be strange to completely forbid using a higher-precision dtype on the right-hand side (unrelated to the array API standard though).

Reply:
What is to do here is to remove the bullet point!  It would be a huge change and those examples would fail *if you would implement it*.

Reply:
That's what I said: "sentence can be removed.".

That said, this thread is about in-place operators, _not_ about `sum`/`mean` (that's higher up - you just ). I'd say:
1. nothing to do standard-wise
2. not high-impact probably without the `sum`/`mean` change
3. it may be considered a design issue, it's pretty weird that `-=` is so different from the out-of-place version. That's exactly what you said higher up:

> For the result-cast safety: I like it! But I need assurance that it isn't annoyingly disruptive, because if it is I don't buy it is worth doing it, especially skipping a deprecation needs a better reason IMO.

Reply:
I think the array API specification is reasonable here - silent downcasting is very risky, certainly for integers. For the purpose of this NEP, maybe we should punt? It means changing the default casting rule for ufuncs from 'same_kind' to 'safe', which may break a lot of things. Maybe it is something that can be experimented with via some environment variable or so (that controls the default).

Reply:
Makes sense - how about I copy the content here into a new issue, and then we can see if it's possible to slowly deprecate?

Reply:
Yes, sounds good.

Reply:
Ah, looked up the original text, and that clarifies a bit more. There are two requirements:

(1) _An in-place operation must not change the data type or shape of the in-place array as a result of [Type Promotion Rules](https://data-apis.org/array-api/latest/API_specification/type_promotion.html#type-promotion) or [Broadcasting](https://data-apis.org/array-api/latest/API_specification/broadcasting.html#broadcasting)._

(2) _An in-place operation must have the same behavior (including special cases) as its respective binary (i.e., two operand, non-assignment) operation._

I was mainly thinking about (1) - "do not upcast" works perfectly fine. It's (2) that is the issue here. So yes, we do deviate a bit. I'll add it to the "not adopted" section then. (2) is perfectly reasonable as a requirement indeed, but there's no hurry here it looks like.

Reply:
I opened gh-25621 for following up on this topic, and added this requirement to "not adopted" now. Resolving the comment. Thanks for working through it @seberg and @mhvk.